modules/tinymce/src/core/main/ts/newline/InsertBlock.ts

```diff
@@ -242,139 +224,198 @@ const addBrToBlockIfNeeded = function (dom, block) {
+      walker.next();
     }
+  }
 
-    // Split the current container block element if enter is pressed inside an empty inner block element
-    if (Settings.shouldEndContainerOnEmptyBlock(editor) && canSplitBlock(dom, containerBlock) && dom.isEmpty(parentBlock)) {
-      // Split container block for example a BLOCKQUOTE at the current blockParent location for example a P
-      newBlock = dom.split(containerBlock, parentBlock);
-    } else {
-      dom.insertAfter(newBlock, parentBlock);
-    }
+  return true;
+};
 
-    NewLineUtils.moveToCaretPosition(editor, newBlock);
+// Used as an internal function below. Dragging it up here to make the function state explicit (and read-only).
+interface InsertNewBlockAfter {
+  containerBlockName: string;
+  newBlockName: string;
+  parentBlockName: string;
+  createNewBlock: (name?: string) => Node;
+  parentBlock: Element;
+  containerBlock: Element;
+  dom: DOMUtils;
+  editor: Editor;
+}
+
+const _insertNewBlockAfter = ({ containerBlockName, newBlockName, parentBlockName, createNewBlock, parentBlock, containerBlock, dom, editor }: InsertNewBlockAfter): Node => {
+  let newBlock: Node;
+
+  // If the caret is at the end of a header we produce a P tag after it similar to Word unless we are in a hgroup
+  if (/^(H[1-6]|PRE|FIGURE)$/.test(parentBlockName) && containerBlockName !== 'HGROUP') {
+    newBlock = createNewBlock(newBlockName);
+  } else {
+    newBlock = createNewBlock();
+  }
+
+  // Split the current container block element if enter is pressed inside an empty inner block element
+  if (Settings.shouldEndContainerOnEmptyBlock(editor) && canSplitBlock(dom, containerBlock) && dom.isEmpty(parentBlock)) {
+    // Split container block for example a BLOCKQUOTE at the current blockParent location for example a P
+    newBlock = dom.split(containerBlock, parentBlock);
+  } else {
+    dom.insertAfter(newBlock, parentBlock);
+  }
+
+  NewLineUtils.moveToCaretPosition(editor, newBlock);
+
+  return newBlock;
+};
+
+const insert = (editor: Editor, evt?: EditorEvent<KeyboardEvent>) => {
+  let newBlock: Node, isAfterLastNodeInContainer = false;
+  const dom = editor.dom;
+  const schema = editor.schema, nonEmptyElementsMap = schema.getNonEmptyElements();
+  const rng = editor.selection.getRng();
+
+  // Creates a new block element by cloning the current one or creating a new one if the name is specified
+  // This function will also copy any text formatting from the parent block and add it to the new one
+  const createNewBlock = function (name?: string) {
+    return _createNewBlock({ container, schema, parentBlockName, newBlockName, dom, parentBlock, editor, editableRoot }, name);
+  };
+
+  // Returns true/false if the caret is at the start/end of the parent block element
+  const isCaretAtStartOrEndOfBlock = function (start?: true) {
+    const output = _isCaretAtStartOrEndOfBlock({ offset, container, isAfterLastNodeInContainer, parentBlock, nonEmptyElementsMap }, start);
```

Comment:
Return the value directly here maybe?

Reply:
Will do, that was a leftover from when I was trying to add logs to debug things